README.md
Puppet Profile Parser
=====================

[![Build Status](https://travis-ci.org/Sharpie/puppet-profile-parser.svg?branch=master)](https://travis-ci.org/Sharpie/puppet-profile-parser)

Tools for parsing profile information from Puppet Server logs and transforming
to various output formats.


Installation
------------

The tool in this repository is the `puppet-profile-parser.rb` script. This
script has no dependencies and can be run from any location where a Ruby
interpreter is present on the `PATH`.

The script requires Ruby 2.0 or newer.

The most recent stable release of the script can be downloaded here:

  [Stable release: 0.3.0][stable-release]


And the latest development version can be downloaded from:

  [Edge release from master branch][edge-release]

  [stable-release]: https://github.com/Sharpie/puppet-profile-parser/releases/download/0.3.0/puppet-profile-parser.rb
  [edge-release]: https://raw.githubusercontent.com/Sharpie/puppet-profile-parser/master/puppet-profile-parser.rb


Usage
-----

### Generating profiles

The `puppet-profile-parser.rb` script scans Puppet Server logs for information
generated when the Puppet profiler is enabled. The profiler is disabled by
default and can be enabled using the [profile setting][profile-setting]
in `puppet.conf`.

  [profile-setting]: https://puppet.com/docs/puppet/5.4/configuration.html#profile

#### Profiling specific agents

A single profile can be generated for an agent by executing a test run
with the `--profile` flag:

    puppet agent -t --profile

Setting `profile=true` in the `[agent]` section of `puppet.conf` and restarting
the `puppet` service will cause all runs to generate profiling information. The
profile results will be located in the Puppet Server logs.

#### Profiling all agents

Profiling can be enabled globally by setting `profile=true` in the `[master]`
section of `puppet.conf` and restarting the `puppetserver` service.

#### Configure Puppet Server logging

Depending on the version in use, there are certain adjustments you'll want to
make to the Puppet Server logging configuration. These adjustments can be made
in the main logback configuration file:

    /etc/puppetlabs/puppetserver/logback.xml

If `puppet --version` is less than `4.8.0`, the level for the `puppetserver`
logger will need to be raised to DEBUG by adding the following configuration
towards the bottom of the file:

    <logger name="debug" level="debug"/>

Or, upgrade the `puppet-agent` package to provide Puppet 4.8.0, where profiling
data is logged at the default INFO level.

Puppet Server should be configured to include the time zone in log timestamps.
This can be done by adjusting the `pattern` of the `F1` appender:

     <pattern>%d{yyyy-MM-dd'T'HH:mm:ss.SSSXXX} %-5p [%t] [%c{2}] %m%n</pattern>

Adding the time zone improves allows logs to be processed with accurate time
stamps.

### Parsing profiles

The `puppet-profile-parser.rb` script expects list of Puppet Server
log files containing `PROFILE` entries:

    ./puppet-profile-parser.rb .../path/to/puppetserver.log [.../more/logs]

The script is capable of reading both plaintext log files and archived
log files that have been compressed with `gzip`. Compressed log files
must have a name that ends in `.gz` in order to be read.

A full list of options can be displayed using the `--help` flag:

    ./puppet-profile-parser.rb --help

### Output formats

The profile parser extracts a "trace" for each request profiled by the Puppet
Server. Within each trace are a number of nested "spans" representing
instrumented oprations executed to generate a response for the request. The
terminology of traces and spans follows the [OpenTracing specification][opentracing-spec].

The parser is capable of rendering traces to `$stdout` in a variety of
output formats which are selected using the `--format` flag. The currently
supported output formats are:

  - Human-readable (default)
  - CSV
  - FlameGraph stacks
  - Zipkin JSON

**NOTE:** Each trace is currently assigned a randomly-generated UUIDv4 as an
identifier. Re-running the script on the same input will result in the same
traces, but with new randomly generated IDs. These IDs are included in the
CSV and Zipkin output formats.

  [opentracing-spec]: https://github.com/opentracing/specification/blob/master/specification.md

#### Human readable

The `human` output format is the default used by the script if the `--format`
flag is not used to select another option This output format displays each
trace parsed from the logs as an indented list followed by summary tables for
function calls, resource evaluations, and "other" operations measured by the
profiler. The summary tables are sorted in terms of "exclusive" time, which is
the time spent on the operation after excluding any time spent on nested child
operations.

In POSIX environments, the traces are also colorized using ANSI color codes.
This behavior can be toggled using the `--color` and `--no-color` flags.


#### CSV

The CSV output format prints a header row followed by a row for each span in
each trace using a comma-separated format. The columns included in the CSV
output are:

  - `timestamp`: ISO-8601 formatted timestamp with timezone indicating when
     the span was logged.

  - `trace_id`: Randomly assigned UUIDv4 for each trace.

  - `span_id`: Dot-delimited sequence of numbers indicating span number
     and nesting depth.

  - `name`: Name of the profiled operation.

  - `exclusive_time_ms`: Milliseconds spent on the span, excluding time
     spent on nested child spans.

  - `inclusive_time_ms`: Milliseconds spent on the span, including time
     spent on nested child spans.


#### FlameGraph

The `flamegraph` output format prints each span in each trace as a semi-colon
delimited call stack followed by the number of milliseconds measured for
that span. This output format can be piped into the `flamegraph.pl` script
from [brendangregg/FlameGraph][flamegraph] to create interactive SVG
visualizations:

    ./puppet-profile-parser.rb -f flamegraph puppetserver.log | \
      path/to/flamegraph.pl --countname ms > puppet_profile.svg

An example SVG generated by `flamegraph.pl` (click for interactive version):

  [![Example FlameGraph, click for interactive version.][example-flamegraph]][example-flamegraph]

  [flamegraph]: https://github.com/brendangregg/FlameGraph
  [example-flamegraph]: https://sharpie.github.io/puppet-profile-parser/assets/puppet_profile.svg


#### Zipkin JSON

The `zipkin` output format prints a JSON array containing each span in each
trace. The JSON array conforms to the `ListOfSpans` data type defined by the
[Zipkin v2 API specification][zipkin-v2-spec]. This allows the JSON output
to be submitted as a POST request to services which implement the API:

    ./puppet-profile-parser.rb -f zipkin puppetserver.log | \
      curl -X POST -H 'Content-Type: application/json' \
        http://<zipkin hostname>:9411/api/v2/spans --data @-

There are two implementations of the Zipkin API that can be spun up quickly
inside of Docker containers:

  - [Jaeger][jaeger], an implementation by Uber now part of the CNCF:

        docker run -d -e COLLECTOR_ZIPKIN_HTTP_PORT=9411 \
          -p 16686:16686 -p 9411:9411 jaegertracing/all-in-one:latest

    Visit <http://localhost:16686> to view profiling data in the Jaeger UI.
    Double check the date range in the "lookback" setting if no traces show up.

  - [Zipkin][zipkin], the original implementation by Twitter:

        docker run -d -p 9411:9411 openzipkin/zipkin

    Visit <http://localhost:9411> to view profiling data in the Zipkin UI.
    Double check the date range in the "lookback" setting if no traces show up.


  [zipkin-v2-spec]: https://zipkin.io/zipkin-api/
  [jaeger]: http://jaeger.readthedocs.io/en/latest/
  [zipkin]: https://zipkin.io/


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.0] - 2018-05-25

### Added

  - A `--debug` flag which enables error backtraces.

  - PuppetDB operations are now parsed explicitly instead of being part
    of the "other" span group.

  - HTTP requests are parsed explicitly and tagged with method, url and
    peer.certname (if available).

  - Support for Ruby 2.0.

### Fixed

  - The script now exits with code 1 if an error is raised.

  - The script now allows for a PROFILE message containing `]` characters.

  - Zipkin output no longer contains a redundant `span.kind` tag.

  - The profile parser exits 1 and prints a clear error message if run
    under a Ruby version older than 2.0.

### Changed

  - The `get` method of the Trace class is now private.

  - The FunctionSpan, ResourceSpan, and OtherSpan classes have been replaced
    by the generic Span class and all parsing logic has been moved to the
    TraceParser class.

  - The CsvOutput, FlameGraphOutput, HumanOutput, and ZipkinOutput classes have
    been nested under a new Formatter class as Formatter::Csv, Formatter:FlameGraph,
    Formatter::Human and Formatter::Zipkin. The `display` method has also been
    re-named to `write`.


## [0.2.0] - 2018-03-18

### Added

  - Colorization of output can be toggled via the `--color` and `--no-color`
    flags.

  - The LogParser class now has a `parse_file` method.

  - Each Trace is assigned a random UUID.

  - CSV output format.

  - JSON output for [v2 of the Zipkin API][zipkin-v2].

  [zipkin-v2]: https://github.com/openzipkin/zipkin-api

### Changed

  - The Namespace class has been re-named to Trace and the Slice classes
    have been renamed to Span. This matches the implementation up with
    [OpenTracing terminology][opentracing-spec].

  - The `profile-parser.rb` script has been re-named to `puppet-profile-parser.rb`
    for clarity.

  - The PuppetProfiler module has been re-named to PuppetProfileParser for
    consistency with the script name.

  - The script prints usage to stderr and exits 1 if no log files are passed.

  [opentracing-spec]: https://github.com/opentracing/specification/blob/master/specification.md


## [0.1.0] - 2018-03-02

### Added

  - RSpec tests.

  - Apache 2 license.

  - Support for reading gzipped log files.

  - Support for FlameGraph output.

### Changed

  - Profiled events are now separated by request and Java thread id. Incomplete
    profiles are dropped.

  - Inclusive and exclusive times are computed for each profile span. Output
    uses exclisive time so that hot spots aren't hidden by spans double
    counting the time taken by their children.

### Removed

  - Dependency on the `colored` and `terminal-table` gems.

  - The `catalog-analyzer.rb` script has been removed. It may return in the
    future, but for now the project will focus on parsing profile data.


## [0.0.1] - 2014-05-19

Initial version by [Adrien Thebo](https://github.com/adrienthebo)


[0.3.0]: https://github.com/Sharpie/puppet-profile-parser/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/Sharpie/puppet-profile-parser/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/Sharpie/puppet-profile-parser/compare/0.0.1...0.1.0
[0.0.1]: https://github.com/Sharpie/puppet-profile-parser/compare/53a9d9f...0.0.1


spec/spec_helper.rb
PROJECT_ROOT = File.expand_path('..', File.dirname(__FILE__)).freeze
SPEC_ROOT = File.expand_path(File.dirname(__FILE__)).freeze

require 'json-schema'

module TestHelpers
  def fixture(name)
    File.join(SPEC_ROOT, 'fixtures', name)
  end
end

RSpec.configure do |config|
  config.include TestHelpers
end


spec/unit/profile_parser_spec.rb
require 'spec_helper'

require "#{PROJECT_ROOT}/puppet-profile-parser.rb"

describe PuppetProfileParser::LogParser do
  describe 'ISO 8601 timestamp parser' do
    subject { PuppetProfileParser::LogParser::ISO_8601 }

    it 'matches timestamps produced by the logback %date pattern' do
      expect(subject.match('2017-09-18 12:01:43,344')).to be_a(MatchData)
    end
  end

  describe 'Puppet Server default logback layout parser' do
    subject { PuppetProfileParser::LogParser::DEFAULT_PARSER }

    it 'matches PROFILE output formatted with the layout' do
      result = subject.match("2018-02-18 18:43:53,501 INFO  [qtp1732817189-1224] [puppetserver] Puppet PROFILE [39776666] 1 Processed request GET /puppet/v3/node/pe-201734-master.puppetdebug.vlan: took 0.1640 seconds\n")

      expect(result).to be_a(MatchData)
      expect(result[:timestamp]).to eq('2018-02-18 18:43:53,501')
      expect(result[:log_level]).to eq('INFO')
      expect(result[:thread_id]).to eq('qtp1732817189-1224')
      expect(result[:java_class]).to eq('puppetserver')
      expect(result[:request_id]).to eq('39776666')
      expect(result[:message]).to eq('1 Processed request GET /puppet/v3/node/pe-201734-master.puppetdebug.vlan: took 0.1640 seconds')
    end
  end

  describe 'when parsing a logfile' do
    subject { described_class.new }
    let(:log_file) { fixture('puppetserver.log') }

    before(:each) { subject.parse_file(log_file) }

    it 'creates a trace for each complete PROFILE with unique ids' do
      expect(subject.traces.flat_map {|t| t.map(&:trace_id)}.uniq.length).to eq(2)
    end

    it 'computes inclusive and exclusive time for each trace' do
      expect(subject.traces.first.inclusive_time).to eq(1500)
      expect(subject.traces.first.exclusive_time).to eq(1250)
    end
  end
end


spec/unit/profile_parser_cli_spec.rb
require 'spec_helper'

require "#{PROJECT_ROOT}/puppet-profile-parser.rb"

describe PuppetProfileParser::CLI do
  # Caputure output from tests.
  # TODO: Add better output control to the CLI.
  original_stdout = $stdout
  let(:output) { StringIO.new }
  before(:each) { $stdout = output }
  after(:each) { $stdout = original_stdout }

  it 'uses file extensions to determine which IO class to use' do
    cli = described_class.new(['foo.log', 'bar.log.gz', 'baz'])

    expect(File).to receive(:open).with('foo.log', any_args).and_return(StringIO.new)
    expect(File).to receive(:open).with('baz', any_args).and_return(StringIO.new)
    expect(Zlib::GzipReader).to receive(:open).with('bar.log.gz').and_return(StringIO.new)

    cli.run
  end
end


profile-parser.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

require 'zlib'
require 'optparse'
require 'securerandom'
require 'csv'
require 'time'
require 'json'
require 'digest/sha2'

module PuppetProfiler
  VERSION = '0.1.0'.freeze

  # Utility functions for terminal interaction
  module Tty
    COLOR_CODES = {
      red: 31,
      green: 32,
      yellow: 33,
    }.freeze

    def self.windows?
      @is_windows ||= (::File::ALT_SEPARATOR == "\\")
    end

    def self.tty?
      @is_tty ||= $stdout.tty?
    end

    def self.colorize(color, string, enable = nil)
      if (!windows?) && ((enable.nil? && tty?) || enable)
        "\033[#{COLOR_CODES[color]}m#{string}\033[0m"
      else
        string
      end
    end

    COLOR_CODES.keys.each do |name|
      define_singleton_method(name) do |string, enable = nil|
        colorize(name, string, enable)
      end
    end
  end

  class Trace
    include Enumerable

    attr_reader :namespace
    attr_reader :object
    attr_reader :trace_id

    # Milliseconds spent on this operation, excluding child operations
    #
    # Equal to inclusive_time minus the sum of inclusive_time for children.
    attr_reader :exclusive_time
    # Milliseconds spent on this operation, including child operations
    attr_reader :inclusive_time
    # Array of operation names
    attr_reader :stack

    def initialize(namespace, object, trace_id = nil)
      @namespace = namespace.split('.')
      @object    = object
      @trace_id  = trace_id || SecureRandom.uuid
      @children  = {}
      @exclusive_time = nil
      @inclusive_time = nil
    end

    def add(ns, object)
      parts = ns.split('.')

      child_ns = parts[0..-2]
      child_id = parts.last

      if(parts == @namespace)
        # We are the object
        @object = object
      elsif @namespace == child_ns
        get(child_id).add(ns, object)
      else
        id = parts.drop(@namespace.size).first
        child = get(id)
        child.add(ns, object)
      end
    end

    def get(id)
      @children[id] ||= Trace.new([@namespace, id].flatten.join('.'),
                                  nil,
                                  @trace_id)
    end

    def each
      yield self

      @children.each do |_, child|
        child.each {|grandchild| yield grandchild }
      end
    end

    # Compute summary statistics
    #
    # This method should be called once all child spans have been added
    # in order to compute summary statistics for the entire trace.
    def finalize!(parent_stack = [], parent_object = nil)
      @stack = parent_stack + [object.name]

      @children.each {|_, child| child.finalize!(@stack, object) }

      @inclusive_time = Integer(object.time * 1000)

      child_time = @children.inject(0) {|sum, (_, child)| sum + child.inclusive_time }
      @exclusive_time = @inclusive_time - child_time
      @exclusive_time = 0 unless (@exclusive_time >= 0)

      # Copy state to our wrapped span.
      object.context[:trace_id] = @trace_id
      object.references << ['child_of', parent_object.id] unless parent_object.nil?
      object.finish!
    end
  end

  class Span
    attr_reader :id
    attr_reader :time
    attr_reader :start_time
    attr_reader :finish_time

    attr_reader :context
    attr_reader :tags
    attr_reader :references

    def initialize(name = nil, finish_time = nil, tags = {})
      # Typically spans are initialized with the start time.
      # But, we're parsing from logs writen _after_ the operation
      # completes. So, its finish time.
      @finish_time = finish_time
      @name = name

      @context = {trace_id: nil, span_id: nil}
      @tags = tags.merge({'component' => 'puppetserver',
                          'span.kind' => 'server'})
      @references = []
    end

    def finish!
      unless @finish_time.nil?
        @start_time = @finish_time - @time
      end
    end
  end

  class FunctionSpan < Span
    attr_reader :function
    alias name function

    def parse(line)
      match = line.match(/([\d\.]+) Called (\S+): took ([\d\.]+) seconds/)
      @id       = match[1]
      @function = match[2]
      @time     = match[3].to_f

      @context[:span_id] = @id

      self
    end

    def inspect
      "function #{@function}"
    end
  end

  class ResourceSpan < Span
    attr_reader :type
    attr_reader :title

    def parse(line)
      match = line.match(/([\d\.]+) Evaluated resource ([\w:]+)\[(.*)\]: took ([\d\.]+) seconds$/)

      @id    = match[1]
      @type  = match[2]
      @title = match[3]
      @time  = match[4].to_f

      @context[:span_id] = @id

      self
    end

    def name
      @name ||= if (@type == 'Class')
                  "#{@type}[#{@title}]"
                else
                  @type
                end
    end

    def inspect
      "resource #{@type}[#{@title}]"
    end
  end

  class OtherSpan < Span
    attr_reader :name

    def parse(line)
      match = line.match(/PROFILE \[\d+\] ([\d\.]+) (.*): took ([\d\.]+) seconds$/)
      @id = match[1]
      @name = match[2]
      @time = match[3].to_f

      @context[:span_id] = @id

      self
    end

    def inspect
      @name
    end
  end

  class TraceParser
    def initialize
      @spans = []
    end

    def parse(line, metadata)
      span_class = case line
                   when /Called/
                     FunctionSpan
                   when /Evaluated resource/
                     ResourceSpan
                   else
                     OtherSpan
                   end

      span = span_class.new(nil, metadata[:timestamp]).parse(line)

      if span.id == '1'
        # We've hit the root of a profile, which gets logged last.
        trace = Trace.new('1', span)

        @spans.each do |child|
          trace.add(child.id, child)
        end

        # Re-set for parsing a new profile and return the completed trace.
        @spans = []

        trace.finalize!
        return trace
      else
        @spans << span

        # Return nil to signal we haven't parsed a complete profile yet.
        return nil
      end
    end
  end

  # Top-level parser for extracting profile data from logs
  class LogParser
    # Regex for parsing ISO 8601 datestamps
    #
    # Basically the regex used by {Time.iso8601} with some extensions.
    #
    # @see https://ruby-doc.org/stdlib-2.4.3/libdoc/time/rdoc/Time.html#method-i-xmlschema
    ISO_8601 = /(?:\d+)-(?:\d\d)-(?:\d\d)
                [T\s]
                (?:\d\d):(?:\d\d):(?:\d\d)
                (?:[\.,]\d+)?
                (?:Z|[+-]\d\d:\d\d)?/ix

    # Regex for parsing Puppet Server logs
    #
    # Matches log lines that use the default logback pattern for Puppet Server:
    #
    #     %d %-5p [%t] [%c{2}] %m%n
    DEFAULT_PARSER = /^\s*
                      (?<timestamp>#{ISO_8601})\s+
                      (?<log_level>[A-Z]+)\s+
                      \[(?<thread_id>\S+)\]\s+
                      \[(?<java_class>\S+)\]\s+
                      (?<message>.*)$/x

    attr_reader :traces

    def initialize
      @traces = []
      @trace_parsers = Hash.new {|h,k| h[k] = TraceParser.new }

      # TODO: Could be configurable. Would be a lot of work to implement
      # a reasonable intersection of Java and Passenger formats.
      #
      # The "PROFILE [<trace_id>] ..." messages also include an id which
      # is set to the Ruby object id created to handle each HTTP request.
      # Could switch to using that instead of the Java thread id.
      @log_parser = DEFAULT_PARSER
    end

    def parse_file(file)
      io = if file.is_a?(IO)
             file
           else
             case File.extname(file)
             when '.gz'
               Zlib::GzipReader.open(file)
             else
               File.open(file, 'r')
             end
           end

      begin
        io.each_line do |line|
          next unless line.match("PROFILE")

          parse_line(line)
        end
      ensure
        io.close
      end
    end

    def parse_line(log_line)
      match = @log_parser.match(log_line)

      if match.nil?
        $stderr.puts("WARN Could not parse log line: #{log_line})")
        return
      end

      # NOTE: The zip can be replaced with match.named_captures, which
      # was added in Ruby 2.4.
      data = match.names.zip(match.captures).map do |k, v|
               if k == "timestamp"
                 # Ruby only allows a subset of the ISO8601 string formats.
                 # Java defaults to printing a format that Ruby doesn't allow.
                 v = Time.iso8601(v.sub(' ', 'T').sub(',','.'))
               end

               [k.to_sym, v]
             end.to_h

      trace_parser = @trace_parsers[data[:thread_id]]
      result = trace_parser.parse(data[:message], data)

      # The TraceParser returns nil unless the log lines parsed so far
      # add up to a complete profile.
      traces << result unless result.nil?
    end
  end

  class CsvOutput
    def initialize(output)
      @output = CSV.new(output)
      @header_written = false
    end

    def display(traces)
      traces.each do |trace|
        trace.each do |span|
          data = convert_span(span.object, span)

          unless @header_written
            @output << data.keys
            @header_written = true
          end

          @output << data.values
        end
      end
    end

    private

    def convert_span(span, trace)
      # NOTE: The Puppet::Util::Profiler library prints seconds with 4 digits
      # of precision, so preserve that in the output.
      #
      # TODO: This outputs in ISO 8601 format, which is great but may not be
      # the best for programs like Excel. Look into this.
      {timestamp: span.start_time.iso8601(4),
       trace_id: span.context[:trace_id],
       span_id: span.context[:span_id],
       name: span.name,
       exclusive_time_ms: trace.exclusive_time,
       inclusive_time_ms: trace.inclusive_time}
    end
  end

  class FlameGraphOutput
    def initialize(output)
      @output = output
    end

    def display(traces)
      traces.each do |trace|
        trace.each do |span|
          span_time = span.exclusive_time

          next if span_time.zero?

          # The FlameGraph script uses ; as a separator for namespace segments.
          span_label = span.stack.map {|l| l.gsub(';', '') }.join(';')

          @output.puts("#{span_label} #{span_time}")
        end
      end
    end
  end

  class HumanOutput
    ELLIPSIS = "\u2026".freeze

    def initialize(output, use_color = nil)
      @output = output
      @use_color = if use_color.nil?
                     output.tty?
                   else
                     use_color
                   end
    end

    def display(traces)
      traces.each do |trace|
        trace.each do |span|
          indent = " " * span.namespace.length
          id = Tty.green(span.object.id, @use_color)
          time = Tty.yellow("(#{span.inclusive_time} ms)", @use_color)

          @output.puts(indent + [id, span.object.inspect, time].join(' '))
        end

        @output.write("\n\n")
      end

      spans = Hash.new {|h,k| h[k] = [] }
      traces.each_with_object(spans) do |trace, span_map|
        trace.each do |span|
          case span.object
          when FunctionSpan
            span_map[:functions] << span
          when ResourceSpan
            span_map[:resources] << span
          when OtherSpan
            span_map[:other] << span
          end
        end
      end

      process_group("Function calls", spans[:functions])
      process_group("Resource evaluations", spans[:resources])
      process_group("Other evaluations", spans[:other])
    end

    private

    def truncate(str, width)
      if (str.length <= width)
        str
      else
        str[0..(width-2)] + ELLIPSIS
      end
    end

    def process_group(title, spans)
      total = 0
      itemized_totals = Hash.new { |h, k| h[k] = 0 }

      spans.each do |span|
        total += span.exclusive_time
        itemized_totals[span.stack.last] += span.exclusive_time
      end

      rows = itemized_totals.to_a.sort { |a, b| b[1] <=> a[1] }

      @output.puts "\n--- #{title} ---"
      @output.puts "Total time: #{total} ms"
      @output.puts "Itemized:"

      # NOTE: Table formatting fixed to 72 columns. Adjusting this based on
      # screen size is possible, but not worth the complexity at this time.
      @output.printf("%-50s | %-19s\n", 'Source', 'Time')
      @output.puts(('-' * 50) + '-+-' + ('-' * 19))
      rows.each do |k, v|
        next if v.zero?

        @output.printf("%-50s | %i ms\n", truncate(k, 50), v)
      end
    end
  end

  class ZipkinOutput
    def initialize(output)
      @output = output
    end

    def display(traces)
      first_loop = true
      @output.write('[')

      traces.each do |trace|
        trace.each do |span|
          next unless (span.inclusive_time > 0)

          if first_loop
            first_loop = false
          else
            @output.write(',')
          end

          @output.write(convert_span(span.object).to_json)
        end
      end

      @output.write(']')
    end

    private

    def convert_span(span)
      # Zipkin requires 16 -- 32 hex characters for trace IDs. We can get that
      # by removing the dashes from a UUID.
      trace_id = span.context[:trace_id].gsub('-', '')
      # And exactly 16 hex characters for span and parent IDs.
      span_id = Digest::SHA2.hexdigest(span.context[:span_id])[0..15]

      result = {"traceId" => trace_id,
                "id" => span_id,
                "name" => span.name,
                "kind" => "SERVER",
                "localEndpoint" => {"serviceName" => "puppetserver"}}

      if (parent = span.references.find {|r| r.first == "child_of"})
        result["parentId"] = Digest::SHA2.hexdigest(parent.last)[0..15]
      end

      # Zipkin reports durations in microseconds and timestamps in microseconds
      # since the UNIX epoch.
      #
      # NOTE: Time#to_i truncates to the nearest second. Using to_f is required
      # for sub-second precision.
      unless span.start_time.nil?
        result["timestamp"] = Integer(span.start_time.to_f * 10**6)
      end
      result["duration"] = Integer(span.time * 10**6)

      result["tags"] = span.tags unless span.tags.empty?

      result
    end
  end

  class CLI
    def initialize(argv = [])
      @log_files = []
      @outputter = nil
      @options = {color: $stdout.tty? }

      @optparser = OptionParser.new do |parser|
        parser.banner = "Usage: puppet-profile-parser [options] puppetserver.log [...]"

        parser.on('-f', '--format FORMAT', String,
                  'Output format to use. One of:',
                  '    human (default)',
                  '    csv',
                  '    flamegraph',
                  '    zipkin') do |format|
          @options[:format] = case format
                              when 'csv', 'human', 'flamegraph', 'zipkin'
                                format.intern
                              else
                                raise ArgumentError, "#{format} is not a supported output format. See --help for details."
                              end
        end

        parser.on('--[no-]color', 'Colorize output.',
                  'Defaults to true if run from an interactive POSIX shell.') do |v|
          @options[:color] = v
        end


        parser.on_tail('-h', '--help', 'Show help') do
          $stdout.puts(parser.help)
          exit 0
        end

        parser.on_tail('--version', 'Show version') do
          puts VERSION
          exit 0
        end
      end

      args = argv.dup
      @optparser.parse!(args)

      # parse! consumes all --flags and their arguments leaving
      # file names behind.
      @log_files += args
      @outputter = case @options[:format]
                   when :csv
                     CsvOutput.new($stdout)
                   when :flamegraph
                     FlameGraphOutput.new($stdout)
                   when :zipkin
                     ZipkinOutput.new($stdout)
                   else
                     HumanOutput.new($stdout, @options[:color])
                   end
    end

    def run
      if @log_files.empty?
        $stdout.puts(@optparser.help)
        exit 0
      end

      parser = LogParser.new

      @log_files.each {|f| parser.parse_file(f)}

      @outputter.display(parser.traces)
    end
  end
end

if File.expand_path(__FILE__) == File.expand_path($PROGRAM_NAME)
  begin
    PuppetProfiler::CLI.new(ARGV).run
  rescue => e
    $stderr.puts("ERROR #{e.class}: #{e.message}")
  end
end
